fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -414,23 +416,27 @@ def reduce(*_: Any) -> None:
                     param.grad.mul_(self.world_size_scaling)
 
                     if self.reduce_fp16:
-                        param.grad.data = param.grad.data.half()
+                        self._grad_fp16_buffer[index] = param.grad.to(torch.float16)
 
                     # Future work includes clearing up the buffer if possible
                     def cleanup() -> None:
                         if dst_rank != self.global_rank:
                             param.grad = None
-                        else:
-                            assert param.grad is not None
-                            param.grad.data = param.grad.data.to(dtype=param.dtype)
+                        elif self._grad_fp16_buffer[index] is not None:
+                            param.grad = self._grad_fp16_buffer[index].to(dtype=param.dtype)  # type: ignore  # mypy is drunk
+
+                        self._grad_fp16_buffer[index] = None
 
                     # Async reduce for this buffer, log the future
                     dst_global_rank = OSS.get_global_rank(self.process_group, dst_rank)
 
                     self._work_handles.append(
                         Workhandle(
                             handle=dist.reduce(
-                                tensor=param.grad.data, dst=dst_global_rank, group=self.process_group, async_op=True
+                                tensor=self._grad_fp16_buffer[index] if self.reduce_fp16 else param.grad,  # type: ignore  # mypy is drunk
```

Comment:
Should this be param.grad.data?

Reply:
I switched to that, AFAIK it should not change much though